pkg/services/extsvcauth/oauthserver/oasimpl/service.go

```diff
@@ -405,3 +407,41 @@ func (*OAuth2ServiceImpl) handleRegistrationPermissions(registration *extsvcauth
 	}
 	return selfPermissions, impersonatePermissions
 }
+
+// handlePluginStateChanged reset the client authorized grant_types according to the plugin state
+func (s *OAuth2ServiceImpl) handlePluginStateChanged(ctx context.Context, event *pluginsettings.PluginStateChangedEvent) error {
+	s.logger.Info("Plugin state changed", "pluginId", event.PluginId, "enabled", event.Enabled)
+
+	// Retrieve client associated to the plugin
+	slug := slugify.Slugify(event.PluginId)
+	client, err := s.sqlstore.GetExternalServiceByName(ctx, slug)
+	if err != nil {
+		if oauthserver.IsErrClientNotFound(err) {
+			s.logger.Debug("No ext svc account with this plugin", "pluginId", event.PluginId)
+			return nil
+		}
+		s.logger.Error("Error fetching service", "pluginId", event.PluginId, "error", err)
+		return err
+	}
+
+	// Since we will change the grants, clear cache entry
+	s.cache.Delete(client.ClientID)
+
+	if !event.Enabled {
+		// Plugin is disabled => remove all grant_types
+		return s.sqlstore.UpdateExternalServiceGrantTypes(ctx, client.ClientID, "")
+	}
+
+	if err := s.setClientUser(ctx, client); err != nil {
+		return err
+	}
+
+	// The plugin declared self permissions (that weren't inherited from impersonation)
+	selfEnabled := len(client.SelfPermissions) > 1 || (len(client.SelfPermissions) == 0 && !(client.SelfPermissions[0].Action == ac.ActionUsersImpersonate))
+	// The plugin declared impersonate permissions
+	impersonateEnabled := len(client.ImpersonatePermissions) > 0
```

Comment:
Computing the available grant types based on (self/impersonate) permissions might not be perfect, but given we will probably remove the OAuth2Server token provider, this is good enough.